File: buckpal/src/main/java/study/architecture/buckpal/account/domain/Account.java
package study.architecture.buckpal.account.domain;

import java.time.LocalDateTime;
import java.util.Optional;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Value;

@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class Account {

  private AccountId id;
  private Money baseLineBalance;
  private ActivityWindow activityWindow;

  public static Account withId(
      AccountId accountId,
      Money baseLineBalance,
      ActivityWindow activityWindow) {
    return new Account(accountId, baseLineBalance, activityWindow);
  }

  public Optional<AccountId> getId() {
    return Optional.ofNullable(id);
  }

  public Money calcualteBalance() {
    return Money.add(
        this.baseLineBalance,
        this.activityWindow.calculateBalance(this.id)
    );
  }

  public boolean withdraw(Money money, AccountId targetAccountId) {
    if (!mayWtitdraw(money)) {
      return false;
    }

    Activity withdrawal = new Activity(
        this.id,
        this.id,
        targetAccountId,
        LocalDateTime.now(),
        money);

    activityWindow.addActivity(withdrawal);
    return true;
  }

  private boolean mayWtitdraw(Money money) {
    return Money.add(this.calcualteBalance(), money.negate()).isPositive();
  }

  public boolean deposit(Money money, AccountId sourceAccountId) {
    Activity deposit = new Activity(
        this.id,
        sourceAccountId,
        this.id,
        LocalDateTime.now(),
        money);

    activityWindow.addActivity(deposit);
    return true;
  }

  @Value
  public static class AccountId {

    private Long value;
  }
}

File: buckpal-kotlin/src/main/kotlin/study/cleanarchitecture/buckpalkotlin/account/domain/Account.kt
package study.cleanarchitecture.buckpalkotlin.account.domain

import java.time.LocalDateTime

class Account(
    /**
     * ID
     */
    private val id: AccountId,
    /**
     * activityWindow 이전의 금액
     */
    private val baselineBalance: Money,
    /**
     * 변경 내역. SlidingWindows 처럼 최근 일정 기간 동안의 내역을 가진다
     */
    private val activityWindow: ActivityWindow,
) {
    /**
     * 잔액 계산
     */
    fun calculateBalance(): Money {
        return Money.add(this.baselineBalance, this.activityWindow.calculateBalance(id))
    }

    /**
     * 출금
     */
    fun withdraw(money: Money, targetAccountId: AccountId): Boolean {
        if (!mayWithdraw(money)) {
            return false
        }

        val withdrawal = Activity(
            ownerAccountId = this.id,
            sourceAccountId = this.id,
            targetAccountId = targetAccountId,
            timestamp = LocalDateTime.now(),
            money = money
        )

        this.activityWindow.addActivity(withdrawal)
        return true
    }

    private fun mayWithdraw(money: Money): Boolean {
        return Money.subtract(this.calculateBalance(), money).isPositive()
    }

    fun deposit(money: Money, sourceAccountId: AccountId): Boolean {
        val deposit = Activity(
            ownerAccountId = this.id,
            sourceAccountId = sourceAccountId,
            targetAccountId = this.id,
            timestamp = LocalDateTime.now(),
            money = money
        )

        this.activityWindow.addActivity(deposit)
        return true
    }
}